src/record_ts.h

```diff
@@ -0,0 +1,873 @@
+     * variables or pointers to variables, for that vcall are stored.
+     * It should not change between invocations and we should therefore be able
+     * to capture it and reuse it when replaying the kernel.
+     */
+    uint32_t capture_call_offset(const void *ptr, size_t dsize) {
+        uint32_t size;
+        size = dsize / type_size[(uint32_t) VarType::UInt64];
+
+        AllocType atype = backend == JitBackend::CUDA ? AllocType::Device
+                                                      : AllocType::HostAsync;
+        uint64_t *data  = (uint64_t *) jitc_malloc(atype, dsize);
+        jitc_memcpy(backend, data, ptr, dsize);
+
+        uint32_t data_var =
+            jitc_var_mem_map(backend, VarType::UInt64, data, size, true);
+
+        RecordVariable rv;
+        rv.ptr   = ptr;
+        rv.state = RecordVarState::Captured;
+        rv.init  = RecordVarInit::Captured;
+        rv.index = data_var;
+
+        uint32_t slot;
+        auto it = this->ptr_to_slot.find(ptr);
+        if (it == this->ptr_to_slot.end()) {
+            slot = this->m_recording.record_variables.size();
+
+            this->m_recording.record_variables.push_back(rv);
+
+            this->ptr_to_slot.insert({ ptr, slot });
+        } else {
+            slot                = it.value();
+            RecordVariable &old = this->m_recording.record_variables[slot];
+            if (old.init != RecordVarInit::None)
+                jitc_fail("record(): Tried to overwrite a initialized variable "
+                          "with an offset buffer!");
+
+            this->m_recording.record_variables[slot] = rv;
+        }
+
+        return slot;
+    }
+
+    /**
+     * This function tries to capture a variable that is not known to the
+     * recording \c ThreadState.
+     * This is unsupported for now and raises an exception.
+     */
+    uint32_t capture_variable(uint32_t index, const void */*ptr*/ = nullptr,
+                              bool /*remember*/ = true, bool test_scope = true,
+                              bool /*overwrite*/ = false) {
+
+        scoped_pause();
+        Variable *v = jitc_var(index);
+        if (v->scope < this->m_internal->scope && test_scope) {
+            jitc_raise(
+                "record(): Variable r%u[%u] -> %p, label=%s, was created "
+                "before recording was started, but it was "
+                "not speciefied as an input variable! This can happen if a "
+                "input type is not fully traversavle, for example when not "
+                "specifying a member in DRJIT_STRUCT, but using it in the "
+                "frozen function.",
+                index, v->size, v->data, jitc_var_label(index));
+        }
+
+        jitc_raise("record(): Variable r%u[%u] -> %p, label=%s, data=%s, of "
+                   "size > 1 was created while recording.",
+                   index, v->size, v->data, jitc_var_label(index),
+                   jitc_var_str(index));
```

Comment:
Is this something the user might encounter? It would be great to have a helpful error message like the one above.